File: _archive/run_migration.py
"""Run SQL migration against Supabase using the PostgREST/Management API."""
import os
import sys
import json
import requests
from dotenv import load_dotenv
from pathlib import Path

load_dotenv(Path(__file__).resolve().parent.parent / ".env")

SUPABASE_URL = os.getenv("SUPABASE_URL")
SERVICE_KEY = os.getenv("SUPABASE_SERVICE_ROLE_KEY")

if not SUPABASE_URL or not SERVICE_KEY:
    print("ERROR: Missing SUPABASE_URL or SUPABASE_SERVICE_ROLE_KEY in .env")
    sys.exit(1)

# Extract project ref from URL
PROJECT_REF = SUPABASE_URL.replace("https://", "").split(".")[0]


def run_sql(sql: str) -> dict:
    """Execute SQL via Supabase's pg REST endpoint (rpc)."""
    # Use the PostgREST rpc endpoint isn't suitable for DDL.
    # Instead, use the management API or the direct postgres connection.
    # Supabase exposes a SQL execution endpoint via the REST API.
    url = f"{SUPABASE_URL}/rest/v1/rpc/"

    # For DDL, we need to use supabase-py or the management API.
    # Let's use the supabase-py client which wraps the API.
    # Actually, the cleanest approach is to use the pg8000 or psycopg2
    # direct connection, or the Supabase Management API.
    #
    # Supabase provides a SQL execution endpoint at:
    # POST https://<project>.supabase.co/pg/query
    # with the service_role key as bearer token

    url = f"{SUPABASE_URL}/pg/query"
    headers = {
        "Authorization": f"Bearer {SERVICE_KEY}",
        "Content-Type": "application/json",
        "X-Supabase-Api-Version": "2024-01-01"
    }
    payload = {"query": sql}

    resp = requests.post(url, headers=headers, json=payload, timeout=30)
    return resp


def run_migration_file(filepath: str):
    """Read and execute a SQL migration file."""
    with open(filepath, "r", encoding="utf-8") as f:
        sql = f.read()

    print(f"Running migration: {filepath}")
    print(f"SQL length: {len(sql)} characters")
    print(f"Target: {SUPABASE_URL}")
    print("-" * 50)

    resp = run_sql(sql)

    if resp.status_code == 200:
        print("Migration completed successfully!")
        try:
            data = resp.json()
            print(f"Response: {json.dumps(data, indent=2)[:500]}")
        except:
            print(f"Response text: {resp.text[:500]}")
    else:
        print(f"Migration failed with status {resp.status_code}")
        print(f"Response: {resp.text[:1000]}")

        # If /pg/query doesn't work, try alternative approach
        if resp.status_code in (404, 405):
            print("\n/pg/query not available, trying alternative SQL execution...")
            return run_migration_via_rpc(sql)

    return resp


def run_migration_via_rpc(sql: str):
    """Fallback: try executing SQL via a custom RPC function or direct REST."""
    # Try the Supabase Management API endpoint
    url = f"https://api.supabase.com/v1/projects/{PROJECT_REF}/database/query"
    headers = {
        "Authorization": f"Bearer {SERVICE_KEY}",
        "Content-Type": "application/json"
    }
    payload = {"query": sql}

    resp = requests.post(url, headers=headers, json=payload, timeout=30)

    if resp.status_code == 200:
        print("Migration completed successfully via Management API!")
        try:
            data = resp.json()
            print(f"Response: {json.dumps(data, indent=2)[:500]}")
        except:
            print(f"Response text: {resp.text[:500]}")
    else:
        print(f"Management API also failed with status {resp.status_code}")
        print(f"Response: {resp.text[:500]}")
        print("\nWill try statement-by-statement execution via PostgREST...")

    return resp


def query_practices():
    """Query the practices table to verify seed data."""
    url = f"{SUPABASE_URL}/rest/v1/practices?select=*"
    headers = {
        "apikey": SERVICE_KEY,
        "Authorization": f"Bearer {SERVICE_KEY}",
        "Content-Type": "application/json"
    }

    resp = requests.get(url, headers=headers, timeout=15)

    if resp.status_code == 200:
        data = resp.json()
        print("\n" + "=" * 50)
        print("PRACTICES TABLE - SEED DATA:")
        print("=" * 50)
        print(json.dumps(data, indent=2))
        return data
    else:
        print(f"\nQuery failed with status {resp.status_code}")
        print(f"Response: {resp.text[:500]}")
        return None


if __name__ == "__main__":
    migration_path = Path(__file__).resolve().parent.parent / "supabase" / "migrations" / "001_initial_schema.sql"

    if not migration_path.exists():
        print(f"ERROR: Migration file not found: {migration_path}")
        sys.exit(1)

    result = run_migration_file(str(migration_path))

    if result and result.status_code == 200:
        print("\nQuerying practices table to verify...")
        query_practices()
    else:
        print("\nMigration may have failed. Attempting to query practices anyway...")
        query_practices()

File: execution/run_migration.mjs
/**
 * Run SQL migration against Supabase via direct PostgreSQL connection.
 * Uses Supavisor (pooler) with JWT auth for the service_role.
 */
import { readFileSync } from "fs";
import { resolve, dirname } from "path";
import { fileURLToPath } from "url";
import { createRequire } from "module";

const require = createRequire(import.meta.url);
const { Client } = require("pg");

const __dirname = dirname(fileURLToPath(import.meta.url));

// Load .env
const envPath = resolve(__dirname, "..", ".env");
const envContent = readFileSync(envPath, "utf-8");
const env = {};
for (const line of envContent.split("\n")) {
  const trimmed = line.trim();
  if (!trimmed || trimmed.startsWith("#")) continue;
  const eqIdx = trimmed.indexOf("=");
  if (eqIdx === -1) continue;
  env[trimmed.slice(0, eqIdx)] = trimmed.slice(eqIdx + 1);
}

const SUPABASE_URL = env.SUPABASE_URL;
const SERVICE_KEY = env.SUPABASE_SERVICE_ROLE_KEY;
const PROJECT_REF = SUPABASE_URL.replace("https://", "").split(".")[0];

// Supabase pooler regions to try
const REGIONS = ["us-east-1", "us-east-2", "us-west-1", "us-west-2", "eu-west-1", "eu-west-2", "eu-central-1", "ap-southeast-1", "ap-northeast-1", "ap-south-1", "sa-east-1"];

async function tryPoolerConnection(region, port) {
  const connString = `postgresql://postgres.${PROJECT_REF}:${SERVICE_KEY}@aws-0-${region}.pooler.supabase.com:${port}/postgres`;
  const client = new Client({
    connectionString: connString,
    ssl: { rejectUnauthorized: false },
    connectionTimeoutMillis: 8000,
  });

  try {
    await client.connect();
    const res = await client.query("SELECT 1 as test");
    console.log(`  Connected! Region: ${region}, Port: ${port}`);
    return client;
  } catch (e) {
    try { await client.end(); } catch (_) {}
    throw e;
  }
}

async function findWorkingConnection() {
  // Try session mode (port 5432) - needed for DDL statements
  // Session mode is required for CREATE TABLE, etc.
  for (const region of REGIONS) {
    console.log(`Trying region: ${region} (session mode, port 5432)...`);
    try {
      return await tryPoolerConnection(region, 5432);
    } catch (e) {
      const msg = e.message || e.toString();
      if (msg.includes("ENOTFOUND") || msg.includes("timeout") || msg.includes("ECONNREFUSED")) {
        // Wrong region, try next
        continue;
      }
      console.log(`  Error: ${msg.slice(0, 120)}`);
      // If auth error, try next region (might be wrong region format)
      if (msg.includes("password") || msg.includes("auth")) continue;
      // For other errors, still try next
      continue;
    }
  }

  // Also try transaction mode (port 6543) as fallback
  for (const region of REGIONS) {
    console.log(`Trying region: ${region} (transaction mode, port 6543)...`);
    try {
      return await tryPoolerConnection(region, 6543);
    } catch (e) {
      continue;
    }
  }

  return null;
}

async function main() {
  console.log(`Project ref: ${PROJECT_REF}`);
  console.log(`Target: ${SUPABASE_URL}`);
  console.log("=".repeat(50));
  console.log("Searching for working Supabase pooler connection...\n");

  const client = await findWorkingConnection();

  if (!client) {
    console.log("\nCould not connect via pooler with JWT auth.");
    console.log("Trying direct connection...");

    // Try direct connection
    const directClient = new Client({
      connectionString: `postgresql://postgres:${SERVICE_KEY}@db.${PROJECT_REF}.supabase.co:5432/postgres`,
      ssl: { rejectUnauthorized: false },
      connectionTimeoutMillis: 8000,
    });

    try {
      await directClient.connect();
      console.log("Connected via direct connection!");
      await runMigration(directClient);
    } catch (e) {
      console.log(`Direct connection failed: ${e.message}`);
      console.log("\n=== CONNECTION FAILED ===");
      console.log("Could not establish a PostgreSQL connection.");
      console.log("Options:");
      console.log("1. Provide database password (from Supabase Dashboard > Settings > Database)");
      console.log("2. Run migration in Supabase Dashboard SQL Editor");
      console.log("3. Run: npx supabase login (then use CLI)");
    }
    return;
  }

  await runMigration(client);
}

async function runMigration(client) {
  const migrationPath = resolve(__dirname, "..", "supabase", "migrations", "001_initial_schema.sql");
  const sql = readFileSync(migrationPath, "utf-8");

  console.log(`\nRunning migration: ${sql.length} characters`);
  console.log("-".repeat(50));

  try {
    await client.query(sql);
    console.log("\nMigration completed successfully!");

    // Verify by querying practices
    const result = await client.query("SELECT * FROM practices");
    console.log("\n" + "=".repeat(50));
    console.log("PRACTICES TABLE - SEED DATA:");
    console.log("=".repeat(50));
    console.log(JSON.stringify(result.rows, null, 2));
  } catch (e) {
    console.error(`\nMigration error: ${e.message}`);
    if (e.message.includes("already exists")) {
      console.log("\nTables may already exist. Checking current state...");
      try {
        const result = await client.query("SELECT * FROM practices");
        console.log(JSON.stringify(result.rows, null, 2));
      } catch (e2) {
        console.error(`Query error: ${e2.message}`);
      }
    }
  } finally {
    await client.end();
  }
}

main().catch(console.error);
